### PR TITLE
#14 Setup redux

### PR DESCRIPTION
**Pull request by jol79**

### CODE REVIEW

**Review by katyakostikova — APPROVED** [review]

**Review by liza-veis — CHANGES_REQUESTED** [review]

**Comment by Liza-Veis** [comment]
There are some changes that you need to do:
* you should use yarn for this project, so just remove package-lock.json and install dependencies via yarn
* now there are a lot of eslint errors, you have to fix them (if you are using vscode, you can install prettier extension)
* you should add .tsx extension only to files with components, all other files must be with .ts extension
* to make the app more scalable let's use the store structure from this [repo](https://github.com/BinaryStudioAcademy/bsa-winter-2022-growup/tree/develop/frontend/src), there is already auth slice, so you can follow it's logic
* you have to wrap App component in storeProvider and pass the store
* develop branch has changed, so you need to do rebase from it and resolve the conflicts

**Comment by jol79** [comment]
Based on the feedback:
- [ ] Use yarn for the project, remove _package-lock.json_ and install dependencies via yarn;
- [ ] Install prettier linter in vs code and refactor errors;
- [ ] Use **.ts** for the files that doesn't contain the component(s);
- [ ] Follow structure of the given repo;
- [ ] Wrap App component in **storeProvider** and pass the store;
- [ ] Rebase branch with last changes on the _develop_.

File: package.json [hunk 1]
```diff
@@ -14,6 +14,7 @@
     "@react-navigation/native": "^6.0.8",
     "@react-navigation/native-stack": "^6.5.0",
     "@reduxjs/toolkit": "^1.7.2",
+    "all": "^0.0.0",
```
Comment (AleksandrTermenzhy) [inline package.json]: ?

File: src/screens/Login/index.tsx [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+import { View } from 'react-native';
+import { Text } from 'react-native-paper';
+import { Provider } from 'react-redux';
+import { SafeAreaView } from 'react-native-safe-area-context';
+import { useAppSelector, useAppDispatch } from '../../hooks/hooks';
```
Comment (AleksandrTermenzhy) [inline src/screens/Login/index.tsx]: Please add blank line to separate libraries imports and in-app

File: src/screens/Login/index.tsx [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import { store } from '../../store/store';
+
+const LoginScreen = () => {
+  return (
+    <Provider store={store}>
```
Comment (AleksandrTermenzhy) [inline src/screens/Login/index.tsx]: What for? You have already wrapped the whole app

File: src/store/user/slice.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import { createSlice, PayloadAction } from '@reduxjs/toolkit';
+
+interface userState {
```
Comment (AleksandrTermenzhy) [inline src/store/user/slice.ts]: Let's agree interfaces the following way: `ISomeInterfaceName`

File: src/hooks/hooks.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
+import type { RootState, AppDispatch } from '../store/store';
+
+export const useAppDispatch = () => useDispatch<AppDispatch>();
+export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
```
Comment (liza-veis) [inline src/hooks/hooks.ts]: Move each hook to a separate file and re-export them from index.ts

File: src/store/user/common.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+enum ActionType {
+  SetUser = 'setUser',
+  RemoveUser = 'removeUser',
```
Comment (liza-veis) [inline src/store/user/common.ts]:
```suggestion
  SIGN_IN = 'auth/sign-in',
```

File: src/store/user/actions.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { createAsyncThunk } from '@reduxjs/toolkit';
+import { signIn, signOut } from './slice';
```
Comment (liza-veis) [inline src/store/user/actions.ts]: We don't need this import here, we won't use slice actions in thunks

File: src/store/user/actions.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+  async (loginPayload: any, { dispatch }) => {
+    dispatch(signIn);
+  }
+);
+
+export const authActions = {
+  signIn,
+  signOut,
+  login,
+};
```
Comment (liza-veis) [inline src/store/user/actions.ts]:
```suggestion
export  { signIn };
```

File: src/store/user/actions.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+import { signIn, signOut } from './slice';
+import { ActionType } from './common';
+
+// TODO: use api call to verify user data
+const login = createAsyncThunk(
+  ActionType.SetUser,
+  async (loginPayload: any, { dispatch }) => {
+    dispatch(signIn);
+  }
+);
```
Comment (liza-veis) [inline src/store/user/actions.ts]:
```suggestion
const signIn = createAsyncThunk(
  ActionType.SIGN_IN,
  async (payload: ISignInPayload) => {
    // TODO: implement login logic
  }
);
```

File: src/store/user/slice.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  isLogged: boolean;
+  isAdmin: boolean;
```
Comment (liza-veis) [inline src/store/user/slice.ts]:
```suggestion
  user: IUser | null;
  isAuthorized: boolean;
```

File: src/store/user/slice.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+  initialState,
+  reducers: {
+    signIn: (state, action: PayloadAction<boolean>) => {
+      state.isLogged = true;
+      state.isAdmin = action.payload;
+    },
+    signOut: state => {
+      state.isLogged = false;
+      state.isAdmin = false;
+    },
```
Comment (liza-veis) [inline src/store/user/slice.ts]: We will use extraReducers and thunk actions for this
```suggestion
  reducers: {},
  extraReducers: builder => {
    builder.addCase(signIn.fulfilled, (state, { payload }) => {
      state.user = payload;
      state.isAuthorized = Boolean(payload);
    });
  },
```

File: src/store/user/slice.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  },
+});
+
+export const { signIn, signOut } = userSlice.actions;
+export const { reducer } = userSlice;
```
Comment (liza-veis) [inline src/store/user/slice.ts]: Import signIn from actions
```suggestion

const { reducer } = slice;

const actions = {
  ...slice.actions,
  signIn,
};

export { reducer, actions };
```

File: src/store/user/slice.ts [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+  isLogged: boolean;
+  isAdmin: boolean;
+}
+
+const initialState: userState = {
+  isLogged: false,
+  isAdmin: false,
+};
+
+export const userSlice = createSlice({
```
Comment (liza-veis) [inline src/store/user/slice.ts]:
```suggestion
const slice = createSlice({
```

File: src/store/user/slice.ts [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+  name: 'user',
```
Comment (liza-veis) [inline src/store/user/slice.ts]: Rename name here and user folder to auth

**Review by liza-veis — CHANGES_REQUESTED** [review]

**Review by katyakostikova — APPROVED** [review]

File: src/App.tsx [hunk 15]
```diff
@@ -2,7 +2,9 @@ import React, { useEffect } from 'react';
 import { NavigationContainer } from '@react-navigation/native';
 import { Provider as PaperProvider } from 'react-native-paper';
 import SplashScreen from 'react-native-splash-screen';
+import { Provider as StoreProvider } from 'react-redux';
 
+import { store } from './store/store';
```
Comment (liza-veis) [inline src/App.tsx]: Let's rename store.ts to index.ts
```suggestion
import { store } from 'src/store';
```

File: src/hooks/appDispatch.ts [hunk 16]
```diff
@@ -0,0 +1,4 @@
+import { useDispatch } from 'react-redux';
+import type { AppDispatch } from '../store/store';
```
Comment (liza-veis) [inline src/hooks/appDispatch.ts]: We don't need to use "type" when we import types. Also we should name files in kebab case, so rename the file to use-app-dispatch.ts
```suggestion
import { useDispatch } from 'react-redux';

import { AppDispatch } from 'src/store';
```

File: src/hooks/appSelector.ts [hunk 17]
```diff
@@ -0,0 +1,4 @@
+import { TypedUseSelectorHook, useSelector } from 'react-redux';
+import type { RootState } from '../store/store';
```
Comment (liza-veis) [inline src/hooks/appSelector.ts]: Same here

File: src/store/actions.ts [hunk 18]
```diff
@@ -0,0 +1 @@
+export { signIn } from './auth';
```
Comment (liza-veis) [inline src/store/actions.ts]: We should re-export here all the actions
```suggestion
export { authActions } from './auth';
```

File: src/store/auth/index.ts [hunk 19]
```diff
@@ -0,0 +1,4 @@
+import { reducer as authReducer } from './slice';
+import { signIn } from './actions';
```
Comment (liza-veis) [inline src/store/auth/index.ts]: authActions already include signIn
```suggestion
import { reducer as authReducer, actions as authActions } from './slice';
```

File: src/store/auth/slice.ts [hunk 20]
```diff
@@ -0,0 +1,35 @@
+import { createSlice, PayloadAction } from '@reduxjs/toolkit';
+
+import { IUser } from 'src/common/types';
+import { signIn } from './actions';
+
+interface IUserState {
```
Comment (liza-veis) [inline src/store/auth/slice.ts]:
```suggestion
interface IAuthState {
```

File: src/store/auth/slice.ts [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+  isAuthorized: boolean;
+}
+
+const initialState: IUserState = {
+  user: null,
+  isAuthorized: false,
+};
+
+const slice = createSlice({
+  name: 'user',
```
Comment (liza-veis) [inline src/store/auth/slice.ts]:
```suggestion
  name: 'auth',
```

File: src/store/store.ts [hunk 22]
```diff
@@ -0,0 +1,10 @@
+import { configureStore } from '@reduxjs/toolkit';
+
+import { rootReducer } from './root-reducer';
+
+export const store = configureStore({
+  reducer: rootReducer,
+});
+
+export type RootState = ReturnType<typeof store.getState>;
+export type AppDispatch = typeof store.dispatch;
```
Comment (liza-veis) [inline src/store/store.ts]: Let's export from this file only store and move these types to types/app
-- app-dispatch.ts
-- root-state.ts
-- index.ts

File: src/screens/Login/index.tsx [hunk 23]
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+import { View } from 'react-native';
+import { Text } from 'react-native-paper';
+import { SafeAreaView } from 'react-native-safe-area-context';
+
+import { useAppDispatch } from 'src/hooks';
+import { useAppSelector } from 'src/hooks';
```
Comment (liza-veis) [inline src/screens/Login/index.tsx]:
```suggestion
import { useAppDispatch , useAppSelector } from 'src/hooks';
```

File: src/screens/Login/index.tsx [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+
+const LoginScreen = () => {
+  const dispatch = useAppDispatch();
+  const isAuthorized = useAppSelector(state => state.auth.isAuthorized);
```
Comment (liza-veis) [inline src/screens/Login/index.tsx]: We will show the login screen only for unathorized users, the screen doesn't need to handle it, so remove this

File: src/store/auth/actions.ts [hunk 25]
```diff
@@ -0,0 +1,17 @@
+import { createAsyncThunk } from '@reduxjs/toolkit';
+import { authApi } from 'src/services';
+
```
Comment (liza-veis) [inline src/store/auth/actions.ts]: Add blank line after library import

**Review by liza-veis — CHANGES_REQUESTED** [review]

**Review by liza-veis — APPROVED** [review]

File: src/screens/Login/index.tsx [hunk 26]
```diff
@@ -0,0 +1 @@
+export { LoginScreen } from './login';
```
Comment (AleksandrTermenzhy) [inline src/screens/Login/index.tsx]: Login screen is out of this PR's scope. Please remove screen files before merge

**Review by AleksandrTermenzhy — APPROVED** [review]

**Pull request merged**